### DISCUSSION TITLE
I want to make both GET and POST api routes with the nuxt3 server, is that possible?

### DISCUSSION THREAD

**Discussion by jhughes-dev** [nuxt/nuxt#17789] [Questions] [answered]
I'm trying to port an existing project that uses Express in NODE.js to use nuxt3. 

I have POST and GET endpoints, and I use the bodyParser from express to handle extracting data. If I create middleware in nuxt3 and try to use the H3 "useBody" it complains everything is not a post request. 

Maybe the current documentation is lacking, but this seems like it's not possible.

**Comment by Diizzayy** [comment, marked as answer]
@jhughes-dev There are two approaches you can take to create POST routes:

- simply calling `useBody` within your [api route](https://v3.nuxtjs.org/docs/directory-structure/server#examples) would inherently make the route accept only POST requests.
```ts
import { useBody} from 'h3'

export default async (req, res) => {
  const body = await useBody(req)

  return 'Hello test'
}
```

- You can create a router and explicitly define your routes.
Given the api route `/server/api/test.ts`
```ts
import { createApp, createRouter } from 'h3'

const app = createApp()
const router = createRouter()

app.use(router)

router.get('/', (req, res) => {
  return 'Hello World'
})

router.post('/demo', (req, res) => {
  res.end('my example')
})

export default app
```

Note that in this example, the name of the file will be the base of the router.
So the route would be accessible at `http://localhost:3000/api/test/demo`

**Reply by jhughes-dev** [reply]
Thanks. That's perfect. 
It looks like you just need to return an app and it works from the api route defined by the filename.

Is is possible to do the same or something similar with middleware? 
I'd like to have a common authentication block that decodes a JWT and passes that along to the next piece of the API.

**Reply by Diizzayy** [reply]
@jhughes-dev From what I understand, You simply need to configure a [server middleware](https://v3.nuxtjs.org/docs/directory-structure/server#server-middleware) which would handle your JWT Authorization.

Note though that, server middlewares run before **every request**, If that isn't the behavior that you want you can use `req.url` to filter out the unnecessary paths.

> and passes that along to the next piece of the API

After you've decoded the JWT in your middleware, you can simply set `req.myInfo = 'whatever'`, doing so will make `myInfo` available in the request context of all requests that the middleware is applied to

@jhughes-dev also, feel free to reach out to me if you need additional assistance

**Reply by mubaidr** [reply]
Is this kind of request handler definiton is still supported or deprecated/ dropped?

**Comment by jhughes-dev** [comment]
This gives me the idea to follow the design for server/client plugins to get selective middleware and single file post/get.

It would look something like this:

server
 | - api
 | --- endpt.post.ts
 | --- endpt.get.ts
 | --- endpt.middleware.ts

Maybe that should be a separate discussion.

**Reply by jhughes-dev** [reply]
Looks like this was implemented a while back, at least in part. I would love to have a middleware or "all" that would be processed for everything within a folder. 
My use case is this:
I have some "public" APIs and some which require Authentication. Currently, I have to manually check if the API call would be routed down one path or another. I would love to have:
|- api
|--- public/
|------endpt.get.ts
|--- private/
|------endpt.get.ts
|------endpt.post.ts
|------middleware.ts
